File: LinkedListHW.cpp
#include <iostream>
#include <vector>
using namespace std;

// Node structure
struct Node {
    int data;
    Node* next;
};

// Helper function to print your list
void print_list(Node* head) {
    Node* current = head;
    while (current != nullptr) {
        cout << current->data << " -> ";
        current = current->next;
    }
    cout << "nullptr" << endl;
}

// Helper function to create a list from a vector of values
Node* create_list(vector<int> values) {
    if (values.empty()) {
        return nullptr;
    }
    
    Node* head = new Node{values[0], nullptr};
    Node* current = head;
    
    for (int i = 1; i < values.size(); i++) {
        current->next = new Node{values[i], nullptr};
        current = current->next;
    }
    return head;
}


// Reverse the linked list
Node* reverse_list(Node* head) {
    // Your code here
    return nullptr;
}


int main() {
    // Test the functions
    Node* head = create_list({10, 20, 30, 40});
    
    cout << "Original list: ";
    print_list(head);
    
    head = reverse_list(head);
    
    cout << "Reversed list: ";
    print_list(head);
    
    // Clean up memory (optional but good practice)
    while (head != nullptr) {
        Node* temp = head;
        head = head->next;
        delete temp;
    }
    
    return 0;
}

File: LinkedListHW.py
# LinkedLists HW
'''
Problem: Write a function that reverses a LinkedList

Your function should take in the head of a list as a parameter, and retun the head of the reversed linked list 
'''

#Starter Ccode below including helper functions 
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

def print_list(head):
    """Helper function to print your list"""
    current = head
    while current != None:
        print(current.data, end=" -> ")
        current = current.next
    print("None")

def create_list(values):
    """Helper function to create a list from a list of values"""
    if not values:
        return None
    head = Node(values[0])
    current = head
    for value in values[1:]:
        current.next = Node(value)
        current = current.next
    return head

def reverse_list(head):
    # Your code here
    prev = None
    current = head
    while current != None:
        next = current.next
        current.next = prev
        prev = current
        current = next
    return prev

list = [1, 2, 4, 8, 16, 32, 64, 128]
head = create_list(list)
print_list(head)

head = reverse_list(head)
print_list(head)